File: public/code/indicators__bb_squeeze.ts
/**
 * Bollinger Squeeze Breakout - TypeScript
 */
import { Candle, Signal, makeSignal, bollingerBands, atr, sma } from "../../common";

export function detectBbSqueezeSignal(
  candles: Candle[],
  opts: {
    bbPeriod?: number; bbStd?: number;
    squeezeLookback?: number; useVolumeFilter?: boolean;
  } = {}
): Signal {
  const { bbPeriod = 20, bbStd = 2, squeezeLookback = 50, useVolumeFilter = true } = opts;
  if (candles.length < bbPeriod + squeezeLookback) return makeSignal({ reason: "Insufficient data" });

  const closes = candles.map((c) => c.close);
  const { upper, middle, lower } = bollingerBands(closes, bbPeriod, bbStd);
  const width = upper.map((u, i) => (u - lower[i]) / middle[i]);
  const current = candles[candles.length - 1];
  const prev = candles[candles.length - 2];
  const i = candles.length - 1;

  const widthRecent = width.slice(-squeezeLookback).filter((v) => !isNaN(v)).sort((a, b) => a - b);
  const q25 = widthRecent[Math.floor(widthRecent.length * 0.25)];
  const wasSqueeze = width.slice(-5, -1).every((v) => v < q25);

  const volumes = candles.map((c) => c.volume);
  const avgVol = sma(volumes, 20)[i];
  const volOk = !useVolumeFilter || current.volume > avgVol * 1.2;

  if (wasSqueeze && current.close > upper[i] && prev.close <= upper[i - 1] && volOk) {
    const entry = current.close;
    const stopLoss = middle[i];
    const risk = entry - stopLoss;
    return makeSignal({
      signal: "long", entry, stop_loss: stopLoss,
      take_profit: [entry + risk * 1.5, entry + risk * 2.5, entry + risk * 4],
      confidence: 0.75, reason: "BB squeeze breakout UP",
      timestamp: current.timestamp, metadata: { width: width[i] },
    });
  }
  if (wasSqueeze && current.close < lower[i] && prev.close >= lower[i - 1] && volOk) {
    const entry = current.close;
    const stopLoss = middle[i];
    const risk = stopLoss - entry;
    return makeSignal({
      signal: "short", entry, stop_loss: stopLoss,
      take_profit: [entry - risk * 1.5, entry - risk * 2.5, entry - risk * 4],
      confidence: 0.75, reason: "BB squeeze breakout DOWN",
      timestamp: current.timestamp, metadata: { width: width[i] },
    });
  }
  return makeSignal({ reason: "No squeeze breakout" });
}

File: public/code/indicators__bb_squeeze.py
"""
Bollinger Squeeze Breakout
==========================
BB bandı daraldığında (squeeze) volatilite düşük, sonra patlama gelir.
Squeeze sonrası fiyatın bant dışına çıkması = entry.

Squeeze: BB genişliği son N mum içindeki en düşük seviyede.
"""
import sys, os
sys.path.append(os.path.join(os.path.dirname(__file__), "..", "..", ".."))
import pandas as pd
import numpy as np
from common import make_signal, bollinger_bands, atr, ema


def detect_bb_squeeze_signal(
    df: pd.DataFrame,
    bb_period: int = 20,
    bb_std: float = 2.0,
    squeeze_lookback: int = 50,
    use_volume_filter: bool = True,
) -> dict:
    if len(df) < bb_period + squeeze_lookback:
        return make_signal(reason="Insufficient data")

    upper, middle, lower = bollinger_bands(df["close"], bb_period, bb_std)
    width = (upper - lower) / middle
    current_atr = atr(df, 14).iloc[-1]
    current = df.iloc[-1]
    prev = df.iloc[-2]

    # Squeeze: son N mumda width en düşük 20%'de
    width_recent = width.iloc[-squeeze_lookback:]
    width_pct = (width.iloc[-1] - width_recent.min()) / (width_recent.max() - width_recent.min() + 1e-10)

    was_in_squeeze = width.iloc[-5:-1].mean() < width_recent.quantile(0.25)

    # Volume kontrolü
    avg_vol = df["volume"].rolling(20).mean().iloc[-1]
    vol_ok = (not use_volume_filter) or (current["volume"] > avg_vol * 1.2)

    # Bullish breakout
    if was_in_squeeze and current["close"] > upper.iloc[-1] and prev["close"] <= upper.iloc[-2] and vol_ok:
        entry = current["close"]
        stop_loss = middle.iloc[-1]
        risk = entry - stop_loss
        return make_signal(
            signal="long",
            entry=entry,
            stop_loss=stop_loss,
            take_profit=[entry + risk * 1.5, entry + risk * 2.5, entry + risk * 4],
            confidence=0.75,
            reason="BB squeeze breakout UP",
            metadata={"width": float(width.iloc[-1])},
        )

    if was_in_squeeze and current["close"] < lower.iloc[-1] and prev["close"] >= lower.iloc[-2] and vol_ok:
        entry = current["close"]
        stop_loss = middle.iloc[-1]
        risk = stop_loss - entry
        return make_signal(
            signal="short",
            entry=entry,
            stop_loss=stop_loss,
            take_profit=[entry - risk * 1.5, entry - risk * 2.5, entry - risk * 4],
            confidence=0.75,
            reason="BB squeeze breakout DOWN",
            metadata={"width": float(width.iloc[-1])},
        )

    return make_signal(reason="No squeeze breakout")
